src/cascadia/TerminalApp/HighlightedTextControl.cpp

```diff
@@ -22,70 +22,150 @@ namespace winrt::TerminalApp::implementation
+        _updateTextAndStyle();
     }
 
-    void HighlightedTextControl::Text(const winrt::TerminalApp::HighlightedText& value)
+    static void _applyStyleToObject(const winrt::Windows::UI::Xaml::Style& style, const winrt::Windows::UI::Xaml::DependencyObject& object)
     {
-        SetValue(_textProperty, winrt::box_value(value));
+        if (!style)
+        {
+            return;
+        }
+
+        static const auto fontWeightProperty{ winrt::Windows::UI::Xaml::Documents::TextElement::FontWeightProperty() };
+
+        const auto setters{ style.Setters() };
+        for (auto&& setterBase : setters)
+        {
+            auto setter = setterBase.as<winrt::Windows::UI::Xaml::Setter>();
+            auto property = setter.Property();
+            auto value = setter.Value();
+
+            if (property == fontWeightProperty) [[unlikely]]
+            {
+                // BODGY - The XAML compiler emits a boxed int32, but the dependency property
+                // here expects a boxed FontWeight (which also requires a u16. heh.)
+                // FontWeight is one of the few properties that is broken like this, and on Run it's the
+                // only one... so we can trivially check this case.
+                const auto weight{ winrt::unbox_value_or<int32_t>(value, static_cast<int32_t>(400)) };
+                value = winrt::box_value(winrt::Windows::UI::Text::FontWeight{ static_cast<uint16_t>(weight) });
+            }
+
+            object.SetValue(property, value);
+        }
     }
 
-    // Method Description:
-    // - This callback is triggered when the Text property is changed. Responsible for updating the view
-    // Arguments:
-    // - o - dependency object that was modified, expected to be an instance of this control
-    // - e - event arguments of the property changed event fired by the event system upon Text property change.
-    // The new value is expected to be an instance of HighlightedText
-    void HighlightedTextControl::_onTextChanged(const DependencyObject& o, const DependencyPropertyChangedEventArgs& e)
+    void HighlightedTextControl::_updateTextAndStyle()
     {
-        const auto control = o.try_as<winrt::TerminalApp::HighlightedTextControl>();
-        const auto highlightedText = e.NewValue().try_as<winrt::TerminalApp::HighlightedText>();
-
-        if (control && highlightedText)
+        const auto textBlock = GetTemplateChild(L"TextView").try_as<winrt::Windows::UI::Xaml::Controls::TextBlock>();
+        if (!textBlock)
         {
-            // Replace all the runs on the TextBlock
-            // Use IsHighlighted to decide if the run should be highlighted.
-            // To do - export the highlighting style into XAML
-            const auto inlinesCollection = control.TextView().Inlines();
-            inlinesCollection.Clear();
+            return;
+        }
+
+        const auto text = Text();
```

Comment:
TODO: we could cache both of these, and only regenerate the runs if nothing changed. that would make it cheaper in the general case. we could also cache the _runs_ and only regenerate the ones that differed, but that's starting to become a lot of work for little gain.